File: bindings/wasm/audit_trail_wasm/examples/src/08_delete_audit_trail.ts
// Copyright 2026 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

/**
 * ## Actors
 *
 * - **Admin client**: Creates the trail and sets up the MaintenanceAdmin role.
 * - **Maintenance admin client**: Holds delete permissions. Attempts (and fails) to delete the
 *   non-empty trail, then batch-deletes all records before removing the trail itself.
 *
 * Demonstrates how to:
 * 1. Show that a non-empty trail cannot be deleted.
 * 2. Empty the trail with deleteBatch.
 * 3. Delete the trail once its records are gone.
 */

import { CapabilityIssueOptions, Data, Permission, PermissionSet } from "@iota/audit-trail/node";
import { strict as assert } from "assert";
import { getFundedClient, TEST_GAS_BUDGET } from "./util";

export async function deleteAuditTrail(): Promise<void> {
    console.log("=== Audit Trail: Delete Trail ===\n");

    // `adminClient` creates the trail and delegates trail maintenance.
    // `maintenanceAdminClient` empties and deletes the trail.
    const adminClient = await getFundedClient();
    const maintenanceAdminClient = await getFundedClient();

    const { output: createdTrail } = await adminClient
        .createTrail()
        .withInitialRecordString("Initial record", "event:created")
        .finish()
        .withGasBudget(TEST_GAS_BUDGET)
        .buildAndExecute(adminClient);

    const trailId = createdTrail.id;

    const maintenanceAdminRole = adminClient.trail(trailId).access().forRole("MaintenanceAdmin");
    await maintenanceAdminRole
        .create(new PermissionSet([Permission.DeleteAllRecords, Permission.DeleteAuditTrail]))
        .withGasBudget(TEST_GAS_BUDGET)
        .buildAndExecute(adminClient);
    await maintenanceAdminRole
        .issueCapability(new CapabilityIssueOptions(maintenanceAdminClient.senderAddress()))
        .withGasBudget(TEST_GAS_BUDGET)
        .buildAndExecute(adminClient);

    const maintenanceTrail = maintenanceAdminClient.trail(trailId);

    // 1. Attempting to delete a non-empty trail should fail.
    let deleteWhileNonEmptySucceeded = false;
    try {
        await maintenanceTrail
            .deleteAuditTrail()
            .withGasBudget(TEST_GAS_BUDGET)
            .buildAndExecute(maintenanceAdminClient);
        deleteWhileNonEmptySucceeded = true;
    } catch {
        // Expected
    }
    assert.equal(deleteWhileNonEmptySucceeded, false, "a trail must be empty before deletion");
    console.log("Deleting the non-empty trail failed as expected.\n");

    // 2. Batch-delete all records.
    const deletedRecords = await maintenanceTrail
        .records()
        .deleteBatch(BigInt(10))
        .withGasBudget(TEST_GAS_BUDGET)
        .buildAndExecute(maintenanceAdminClient);
    console.log("Deleted", deletedRecords.output, "record(s) before trail removal.\n");

    const count = await maintenanceTrail.records().recordCount();
    assert.equal(count, 0n, "trail should have no records after batch delete");

    // 3. Delete the now-empty trail.
    const deletedTrail = await maintenanceTrail
        .deleteAuditTrail()
        .withGasBudget(TEST_GAS_BUDGET)
        .buildAndExecute(maintenanceAdminClient);
    console.log("Trail deleted:");
    console.log("  trail_id =", deletedTrail.output.trailId);
    console.log("  timestamp =", deletedTrail.output.timestamp);

    let getAfterDeleteSucceeded = false;
    try {
        await maintenanceTrail.get();
        getAfterDeleteSucceeded = true;
    } catch {
        // Expected
    }
    assert.equal(getAfterDeleteSucceeded, false, "deleted trail should no longer be readable");
}

File: examples/audit-trail/08_delete_audit_trail.rs
// Copyright 2020-2026 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

//! ## Actors
//!
//! - **Admin client**: Creates the trail and sets up the MaintenanceAdmin role.
//! - **Maintenance admin client**: Holds delete permissions. Attempts (and fails) to delete the non-empty trail, then
//!   batch-deletes all records before removing the trail itself.

use std::collections::HashSet;

use anyhow::{Result, ensure};
use audit_trail::core::types::{CapabilityIssueOptions, Data, InitialRecord, Permission, PermissionSet};
use examples::get_funded_audit_trail_client;
use product_common::core_client::CoreClient;

/// Demonstrates how to:
/// 1. Show that a non-empty trail cannot be deleted.
/// 2. Empty the trail with `delete_records_batch`.
/// 3. Delete the trail once its records are gone.
#[tokio::main]
async fn main() -> Result<()> {
    println!("=== Audit Trail: Delete Trail ===\n");

    // Use a maintenance client to keep deletion permissions separate from trail creation.
    let admin_client = get_funded_audit_trail_client().await?;
    let maintenance_admin_client = get_funded_audit_trail_client().await?;

    let created_trail = admin_client
        .create_trail()
        .with_initial_record(InitialRecord::new(
            Data::text("Initial record"),
            Some("event:created".to_string()),
            None,
        ))
        .finish()
        .build_and_execute(&admin_client)
        .await?
        .output;

    let trail_id = created_trail.trail_id;
    let maintenance_admin_role = "MaintenanceAdmin";
    let admin_trail = admin_client.trail(trail_id);

    // The Admin capability authorizes the maintenance role and capability delegation.
    admin_trail
        .access()
        .for_role(maintenance_admin_role)
        .create(
            PermissionSet {
                permissions: HashSet::from([Permission::DeleteAllRecords, Permission::DeleteAuditTrail]),
            },
            None,
        )
        .build_and_execute(&admin_client)
        .await?;
    admin_trail
        .access()
        .for_role(maintenance_admin_role)
        .issue_capability(CapabilityIssueOptions {
            issued_to: Some(maintenance_admin_client.sender_address()),
            valid_from_ms: None,
            valid_until_ms: None,
        })
        .build_and_execute(&admin_client)
        .await?;

    let maintenance_trail = maintenance_admin_client.trail(trail_id);

    let delete_while_non_empty = maintenance_trail
        .delete_audit_trail()
        .build_and_execute(&maintenance_admin_client)
        .await;
    ensure!(delete_while_non_empty.is_err(), "a trail must be empty before deletion");
    println!("Deleting the non-empty trail failed as expected.\n");

    // Batch delete skips locked records and returns the deleted sequence numbers before trail deletion.
    let deleted_records = maintenance_trail
        .records()
        .delete_records_batch(10)
        .build_and_execute(&maintenance_admin_client)
        .await?
        .output;
    println!("Deleted record sequence numbers {deleted_records:?} before trail removal.\n");

    ensure!(maintenance_trail.records().record_count().await? == 0);

    let deleted_trail = maintenance_trail
        .delete_audit_trail()
        .build_and_execute(&maintenance_admin_client)
        .await?
        .output;
    println!(
        "Trail deleted:\n  trail_id = {}\n  timestamp = {}",
        deleted_trail.trail_id, deleted_trail.timestamp
    );

    ensure!(
        maintenance_trail.get().await.is_err(),
        "deleted trail should no longer be readable"
    );

    Ok(())
}
